### PR TITLE
os/bluestore/BlueFS: add bluefs_sync_write option

### PR DESCRIPTION
**Pull request by liewegas**

### CODE REVIEW

File: src/os/bluestore/BlockDevice.h [hunk 1]
```diff
@@ -107,6 +107,10 @@ class BlockDevice {
     uint64_t len,
     char *buf,
     bool buffered) = 0;
+  virtual int write(
```
Comment (ifed01) [inline src/os/bluestore/BlockDevice.h]: This will break NVMEDevice build if any. Need implementation there too

File: src/os/bluestore/BlueFS.cc [hunk 2]
```diff
@@ -1740,13 +1750,19 @@ int BlueFS::_fsync(FileWriter *h, std::unique_lock<std::mutex>& l)
...
   uint64_t old_dirty_seq = h->file->dirty_seq;
 
-  list<FS::aio_t> completed_ios;
-  _claim_completed_aios(h, &completed_ios);
-  lock.unlock();
-  wait_for_aio(h);
-  completed_ios.clear();
-  flush_bdev();
-  lock.lock();
+  if (!cct->_conf->bluefs_sync_write) {
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc]: copy-paste of a block in _flush_and_sync_log, make a single func?

File: src/os/bluestore/KernelDevice.cc [hunk 3]
```diff
@@ -525,6 +525,60 @@ void KernelDevice::aio_submit(IOContext *ioc)
   }
 }
 
+int KernelDevice::write(
```
Comment (ifed01) [inline src/os/bluestore/KernelDevice.cc]: I'd suggest to reuse the sync path from aio_write code as much as possible here.... may be make a standalone internal func?

**Comment by liewegas** [comment]
retest this please

**Comment by tchaikov** [comment]
needs rebase

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-05-05_19:23:54-rados-wip-sage-testing---basic-smithi/

**Pull request merged**